File: DataStructures/Tree/FenwickTree/c/FenwickTree.c
// Sum of Given range Implementation

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

int getSum(int f[],int index){

	int prefixSum = 0;
	while(index > 0){
		prefixSum += f[index];
		index -= (index & (-index));
	}
	return prefixSum;
}
void update(int f[],int n,int index,int val){
	while(index <= n){
		f[index] += val;
		index += (index & (-index));
	}
}
void build(int fenwick[], int a[],int n){

	for(int i=1;i<=n;i++)
		fenwick[i]=0;
	for(int i=0;i<n;i++)
		update(fenwick,n,i+1,a[i]);
}

int main(){

	FILE* fp = freopen("../input.txt", "r", stdin);
	freopen("../output.txt", "w", stdout);

	int i,n, a[n];
	scanf("%d", &n);
	for(i=0; i<n; i++)
		scanf("%d", &a[i]);

	int fenwick[n+1];
	build(fenwick, a, n);

	while(!feof(fp)){

		char opt[20]; 
		fgets(opt, 20, fp); // it reads string from a file, one line at a time

		if(!strncmp(opt,"sum",strlen("sum"))){ 
			char choice[10];
			int l,r;
			sscanf(opt, "%s %d %d" , choice, &l, &r); //sscanf reads formatted input from string

			if(l < 0 || r > n-1 || l > r){
				printf("Invalid Input\n");
			}
			else{
				int sum = getSum(fenwick, r+1) - getSum(fenwick, l);
				printf("sum in the range (%d, %d) is %d\n",l,r,sum);
			}
		}
		else if(!strncmp(opt,"update",strlen("update"))){
			char choice[10];
			int index, newval;
			sscanf(opt, "%s %d %d" , choice, &index, &newval);

			int diff = newval-a[index];
			a[index] = newval;
			update(fenwick,n,index+1,diff); 
		}
	}
	fclose(fp); 
}

File: DataStructures/Tree/FenwickTree/python/FenwickTree.py
# Sum of given range implementation

import sys
import math

def getSum(f, index):

	prefixSum = 0
	while(index > 0):
		prefixSum += f[index]
		index -= (index & (-index))
	return prefixSum

def update(f, n, index, val):
	while(index <= n):
		f[index] += val
		index += (index & (-index))

def build(a, n):

	fenwick = [0] * (n+1)
	for i in range(0, n):
		update(fenwick,n,i+1,a[i]);
	return fenwick

if __name__ == "__main__":

	sys.stdin = open('../input.txt', 'r')
	sys.stdout = open('../output.txt', 'w')

	n = int(sys.stdin.readline())
	a = sys.stdin.readline().split(" ") # [:-1] to eliminate reading of newline
	a = [int(i) for i in a]

	f = build(a, n)

	for line in sys.stdin:

		if line[-1] == '\n':
			opt = line[:-1].split(" ")
		else:
			opt = line.split(" ")
		
		if opt[0] == "sum":
			l, r = int(opt[1]), int(opt[2])
			if(l < 0  or r > n-1 or l > r):
				print("Invalid input ", end = "\n")
			else:
				sum_ = getSum(f, r+1) - getSum(f, l)
				print("sum in the range (",l,", ",r,") is ", sum_, end = "\n")
		elif opt[0] == "update":
			index, newval = int(opt[1]), int(opt[2])
			diff = newval-a[index];
			a[index] = newval;
			update(f, n, index+1, diff);